### PR TITLE
Introduce mode: stack-monitoring for Metricbeat stack modules

### PR DESCRIPTION
**Pull request by ycombinator**
## What does this PR do?

Introduces a new `mode` setting for Metricbeat Elastic Stack modules. The valid values for this setting are `default` or `stack-monitoring`. This setting replaces the old `xpack.enabled` setting, which is still available but deprecated after this PR. 

Users should convert from using the deprecated setting to the new setting as shown below.

| Old setting (now deprecated) | New setting |
| --- | --- |
| `xpack.enabled: false` (default value) | `mode: default` (default value) |
| `xpack.enabled: true` | `mode: stack-monitoring` |

Corresponding to the setting change / deprecation, users should also use the new module configuration variant name as shown below.

| Old way (now deprecated) | New way |
| --- | --- |
| `metricbeat modules enable elasticsearch-xpack` | `metricbeat modules enable elasticsearch-stack-monitoring` |
| `metricbeat modules disable elasticsearch-xpack` | `metricbeat modules disable elasticsearch-stack-monitoring` |
| `metricbeat modules enable kibana-xpack` | `metricbeat modules enable kibana-stack-monitoring` |
| `metricbeat modules disable kibana-xpack` | `metricbeat modules disable kibana-stack-monitoring` |
| `metricbeat modules enable logstash-xpack` | `metricbeat modules enable logstash-stack-monitoring` |
| `metricbeat modules disable logstash-xpack` | `metricbeat modules disable logstash-stack-monitoring` |
| `metricbeat modules enable beat-xpack` | `metricbeat modules enable beat-stack-monitoring` |
| `metricbeat modules disable beat-xpack` | `metricbeat modules disable beat-stack-monitoring` |

## Why is it important?

A couple of reasons:
- We are trying to get rid of the term `xpack` in as many places as possible, and
- The current setting name (`xpack.enabled`) and module configuration variant suffix (`-xpack`) are not very descriptive of their purpose.

## What does this PR NOT do?

As mentioned above, one of the motivations for the `mode` setting is to get rid of the term `xpack`. This term is still used in code in identifiers. This PR does not update those uses. That will be done in a follow up PR, mainly to keep the scope and size of this PR small.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues
- elastic/beats#9424

### CODE REVIEW

File: metricbeat/module/logstash/logstash_test.go [hunk 1]
```diff
@@ -65,11 +65,30 @@ func TestGetVertexClusterUUID(t *testing.T) {
 	}
 }
 
-func TestXPackEnabledMetricSets(t *testing.T) {
+func TestXPackEnabledMetricsets(t *testing.T) {
```
Comment (ycombinator) [inline metricbeat/module/logstash/logstash_test.go:68]: Just making this consistent with other stack modules' tests.

File: metricbeat/module/logstash/metricset.go [hunk 2]
```diff
@@ -15,16 +15,37 @@
...
+	if err != nil {
+		return nil, err
 	}
+
+	return &MetricSet{
+		base,
+		http,
+		config.GetMode() == elastic.ModeStackMonitoring,
+	}, nil
 }
```
Comment (ycombinator) [inline metricbeat/module/logstash/metricset.go:51]: I have no idea why GitHub thinks this is a rename! I deleted the file on the left completely and created the file on the right completely from scrath.

File: metricbeat/helper/elastic/mode.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+// Unpack unmarshals a Mode string to a Mode. This implements
+// ucfg.StringUnpacker.
+func (m *Mode) Unpack(str string) error {
+	switch strings.ToLower(str) {
+	case "default":
+		*m = ModeDefault
+	case "stack-monitoring":
+		*m = ModeStackMonitoring
+	default:
+		return fmt.Errorf("unknown mode: %v", str)
```
Comment (urso) [inline metricbeat/helper/elastic/mode.go:47]: errors on Unpack are user-visible as the beat start might fail if the config can not be applied (it's some kind of config validation). We should try to guide the user a little in the error message what needs to be done.

Proposal:
`s/unknown mode: %v/mode %v not supported, must be one of default or stack-monitoring/`

Comment (ycombinator) [reply metricbeat/helper/elastic/mode.go]: Fixed in 16a503ef5.

**Comment by ycombinator** [comment]
We're not sure if the `xpack.enabled` flag is going to be around in 8.0.0 / `master`. So closing this PR unmerged for now.

**Pull request closed without merging**